[Serialization] Follow typealiases when nominals are renamed.

jrose-apple opened the pull request:
When an Objective-C type is renamed, the importer leaves behind a typealias marked "unavailable". Look through that when we're expecting a nominal type or bound generic type.

There's still more work to do here: in the case of a typealias *not* left behind by the importer, there's a chance the generic arguments of the typealias aren't just directly forwarded to the underlying nominal. The right thing to do here is to build a substitution map and call `subst()` on the underlying type.

jrose-apple committed:
[Serialization] Look through unexpected typealiases.

When an Objective-C type is renamed, the importer leaves behind a
typealias marked "unavailable". Look through that when we're expecting
a nominal type or bound generic type.

There's still more work to do here: in the case of a typealias /not/
left behind by the importer, there's a chance the generic arguments
of the typealias aren't just directly forwarded to the underlying
nominal. The right thing to do here is to build a substitution map
and call subst() on the underlying type.

jrose-apple committed:
[Serialization] Find renamed protocols via typealiases.

If a type declares conformance to a protocol that is then renamed,
deserialization should be able to find the protocol.

jrose-apple commented:
@swift-ci Please test

jrose-apple committed:
[Serialization] Handle cases where a conforming type has been renamed.

...and there was a cross-reference to its conformance. This is a simple
extension of the previous commit.

At this point, there should be no cast<>s in Deserialization.cpp that
aren't guaranteed structurally in some way.

jrose-apple commented:
@swift-ci Please test

DougGregor commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3875,9 +3884,15 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
+        typeOrOffset = alias->getDeclaredInterfaceType();
+      break;
+    }
+    auto nominal = cast<NominalTypeDecl>(decl);
```
Why isn't this another use of `castIgnoringSugar`? Does it need to be more strict?

jrose-apple replied:
The code immediately above handles the typealias case, and tries to decide whether to preserve sugar.

(I kept doing this too: "hey, that's a `cast` I missed! oh, no, that's what I literally just handled".)

DougGregor replied:
Ah, okay.

DougGregor reviewed (approved):
Very nice!

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4148,7 +4163,21 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
+      auto *nominal = alias->getUnderlyingTypeLoc().getType()->getAnyNominal();
+      assert(nominal && "alias does not resolve to a nominal type");
+      // This is not a very good test because of `typealias Foo<X> = Bar<X?>`,
+      // but it's better than nothing.
```
Did you try comparing interface types as well as signatures?

jrose-apple replied:
Oh, of course, the canonical interface types should also match up. Then we're back to your X/Y, Y/X case. Thanks.

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4148,7 +4163,21 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
+    if (auto *alias = dyn_cast<TypeAliasDecl>(decl)) {
+      // FIXME: We should substitute through the alias rather than just relying
+      // on it forwarding arguments perfectly to the nominal.
+      auto *nominal = alias->getUnderlyingTypeLoc().getType()->getAnyNominal();
```
I think `alias->getDeclaredInterfaceType()->getAnyNominal()` is better. It should be equivalent.

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4428,7 +4457,7 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
                                                       genericID, parentID);
 
-    auto genericDecl = cast<NominalTypeDecl>(getDecl(genericID));
+    auto genericDecl = cast<GenericTypeDecl>(getDecl(genericID));
```
Oh, cute. This is a bug fix. Here's a test case:

```
Slavas-MBP:swift slava$ cat ta.swift
import other

func f(_: GG<Int>) {}
Slavas-MBP:swift slava$ cat other.swift
public struct Outer { public typealias G<T> = T }

public typealias GG = Outer.G
```

Building other.swift by itself without WMO is enough to crash because we deserialize the unbound generic type in merge-modules:

```
Slavas-MBP:swift slava$ ../build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc other.swift -emit-module
Assertion failed: (isa<X>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file /Users/slava/new/llvm/include/llvm/Support/Casting.h, line 236.
0  swift                    0x00000001089da068 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x00000001089da716 SignalHandler(int) + 454
2  libsystem_platform.dylib 0x00007fff8a50abba _sigtramp + 26
3  swift                    0x000000010627149a swift::ModuleFile::getGenericSignatureOrEnvironment(llvm::PointerEmbeddedInt<unsigned int, 31>, bool) + 1450
4  libsystem_c.dylib        0x00007fff8a391420 abort + 129
5  libsystem_c.dylib        0x00007fff8a358893 basename_r + 0
6  swift                    0x0000000106278883 swift::ModuleFile::getTypeChecked(llvm::PointerEmbeddedInt<unsigned int, 31>) + 9411
7  swift                    0x0000000106266130 swift::ModuleFile::getType(llvm::PointerEmbeddedInt<unsigned int, 31>) + 32
8  swift                    0x000000010626a092 swift::ModuleFile::getDeclChecked(llvm::PointerEmbeddedInt<unsigned int, 31>, llvm::Optional<swift::DeclContext*>) + 16066
9  swift                    0x000000010626530e swift::ModuleFile::getDecl(llvm::PointerEmbeddedInt<unsigned int, 31>, llvm::Optional<swift::DeclContext*>) + 78
10 swift                    0x00000001062a1b68 swift::ModuleFile::getTopLevelDecls(llvm::SmallVectorImpl<swift::Decl*>&) + 920
11 swift                    0x000000010668279f swift::ModuleDecl::getTopLevelDecls(llvm::SmallVectorImpl<swift::Decl*>&) const + 63
12 swift                    0x0000000105e6f38e swift::ClassHierarchyAnalysis::init() + 78
13 swift                    0x0000000105e62d39 swift::createClassHierarchyAnalysis(swift::SILModule*) + 153
14 swift                    0x0000000105f2a73c swift::SILPassManager::SILPassManager(swift::SILModule*, llvm::StringRef) + 460
15 swift                    0x0000000105f361a7 swift::runSILDiagnosticPasses(swift::SILModule&) + 135
16 swift                    0x00000001056b5ca8 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 12952
17 swift                    0x0000000105670280 main + 3312
18 libdyld.dylib            0x00007fff8a2fd255 start + 1
```

If you build it with WMO on, other.swift builds, but you'll get the same crash when you build ta.swift.

Do you mind adding this as as test case somewhere?

jrose-apple replied:
Thanks! I couldn't think of a test case because I didn't know where UnboundGenericTypes actually got serialized. I'll split it out into its own commit and use your test case.

jrose-apple committed:
[Serialization] Handle UnboundGenericTypes of generic typealiases.

Simple latent asserts-only bug. Thanks to Slava for the test case.

jrose-apple committed:
[ASTVerifier] Add missing verifyCheckedBase for GenericTypeDecl.

All tests still pass, so it looks like we haven't regressed.

jrose-apple committed:
[ClangImporter] Use UnboundGenericTypes in compatibility typealiases.

When a type is renamed, we leave behind a "compatibility typealias"
whose underlying type uses the new name. For generic types, though, we
were using the generic parameters and environment of the original
type, which is completely bogus. "Fix" this by just dropping the
generic part entirely and making a typealias that refers to the
/unbound/ generic type, as if written as `typealias OldName = NewName`
instead of `typealias OldName<Element> = NewName<Element>`. The rest
of the compiler can handle that fine.

jrose-apple committed:
[ClangImporter] Add tests for generic compatibility typealiases.

The implementation here is pretty slapdash, and I'm about to do
something safer, but it seems to be working now and I don't want
to break it.

jrose-apple commented:
Feedback incorporated, and now built on #8751.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
Realized this isn't safe in the general case because the symbol names of functions *trafficking* in these types needs to stay consistent. Hm.

slavapestov commented:
@jrose-apple Yeah, it seems to me this is only useful for imported decls anyway. Giving them their own mangling based on their C names sounds like a great idea.

jrose-apple commented:
Closing for now so it doesn't get accidentally merged. I'll keep the branch around for now though.
